querygen/querygen_test.go

```diff
@@ -0,0 +1,151 @@
+		`objectRef.apiVersion=mvindex('objectRef.apiVersion', 0),` +
+		`objectRef.name=mvindex('objectRef.name', 0),` +
+		`verb=mvindex('verb', 0),` +
+		`requestReceivedTimestamp=mvindex('requestReceivedTimestamp', 0)`
+	out := GenDedupEval(fields)
+	if out != expected {
+		t.Errorf("GenDedupEval() returns:\n  %q, expected\n  %q", out, expected)
+	}
+}
+
+func TestGenApplicationQuery(t *testing.T) {
+	fields := []string{
+		"auditID",
+		"impersonatedUser.username",
+		"user.username",
+		"objectRef.resource",
+		"objectRef.namespace",
+		"objectRef.apiGroup",
+		"objectRef.apiVersion",
+		"objectRef.name",
+		"verb",
+		"requestReceivedTimestamp",
+	}
+	json_properties := map[string]string{
+		"apiGroup":   "objectRef.apiGroup",
+		"apiVersion": "objectRef.apiVersion",
+		"kind":       "objectRef.resource",
+		"name":       "objectRef.name",
+	}
+	expected := `search ` +
+		`index="some_index" ` +
+		`log_type=audit ` +
+		`NOT verb IN (get, watch, list, deletecollection) ` +
+		`"responseStatus.code" IN (200, 201) ` +
+		`"objectRef.apiGroup"="appstudio.redhat.com" ` +
+		`"objectRef.resource"="applications" ` +
+		`("impersonatedUser.username"="*" OR (user.username="*" AND NOT user.username="system:*")) ` +
+		`(verb!=create OR "responseObject.metadata.resourceVersion"="*")` +
+		`|` + GenDedupEval(fields) +
+		`|` + GenTrackFields(TrackFieldSpec{true, true, true}, json_properties)
+	out := GenApplicationQuery("some_index")
+	if out != expected {
+		t.Errorf("GenApplicationQuery() = %v, want %v", out, expected)
+	}
+}
+
+func TestGenPropertiesJSONExpr(t *testing.T) {
+	type args struct {
+		properties_map map[string]string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: "Application query properties",
+			args: args{properties_map: map[string]string{
+				"apiGroup":   "objectRef.apiGroup",
+				"apiVersion": "objectRef.apiVersion",
+				"kind":       "objectRef.resource",
+				"name":       "objectRef.name",
+			}},
+			want: `json_object(` +
+				`"apiGroup",'objectRef.apiGroup',` +
+				`"apiVersion",'objectRef.apiVersion',` +
+				`"kind",'objectRef.resource',` +
+				`"name",'objectRef.name'` +
+				`)`,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := GenPropertiesJSONExpr(tt.args.properties_map); got != tt.want {
+				t.Errorf("GenPropertiesJSONExpr() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestGenTrackFields(t *testing.T) {
+	json_properties := map[string]string{
+		"apiGroup":   "objectRef.apiGroup",
+		"apiVersion": "objectRef.apiVersion",
+		"kind":       "objectRef.resource",
+		"name":       "objectRef.name",
+	}
+	type args struct {
+		spec           TrackFieldSpec
+		properties_map map[string]string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: "with userid, verb, subject & app props",
+			args: args{
+				spec: TrackFieldSpec{
+					with_userid:     true,
+					with_ev_verb:    true,
```

Comment:
We have a parameterize test, but don't we have only the case in which all 3 booleans are true?

Reply:
well, I was putting this in place for the time I add an actual query with different values, but since it took long enough to get to where we are, I pushed that off to the future.

Reply:
can we resolve this?